File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:

        def traverse(node, find, arr):
            arr.append(node)
            if node.val == find.val:
                return arr
            if node.val > find.val:
                return traverse(node.left, find, arr)
            else:
                return traverse(node.right, find, arr)
        
        p_arr = traverse(root, p, [])
        q_arr = traverse(root, q, [])
        res = None
        for i in range(min(len(p_arr), len(q_arr))):
            if p_arr[i] == q_arr[i]:
                res = p_arr[i]
            
        return res



        

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    List<Integer> arr;

    private void bfs(TreeNode node) {
        if (node== null) {
            return;
        }
        bfs(node.left);
        arr.add(node.val);
        bfs(node.right);
    }
    public int kthSmallest(TreeNode root, int k) {
        arr = new ArrayList<>();
        bfs(root);

        return arr.get(k-1);
    }
}
